pkg/graphviz/traceflow.go

```diff
@@ -0,0 +1,246 @@
+	v1.FORWARDED: "Forwarded",
+	v1.DROPPED:   "Dropped",
+}
+
+func genSubGraph(graph *cgraph.Graph, firstNodeName string, obs []v1.Observation, dir cgraph.DirType, addNodeNum int) []*cgraph.Node {
+	var nodes []*cgraph.Node
+
+	// Construct the first node. Show it only if we know the name of it.
+	node, _ := graph.CreateNode(firstNodeName)
+	nodes = append(nodes, node)
+	if len(firstNodeName) > 0 {
+		node.SetFontColor("white")
+		node.SetColor("royalblue1")
+		node.SetStyle(cgraph.FilledNodeStyle)
+	} else {
+		node.SetStyle("invis")
+	}
+
+	// Reorder the observations according to the direction of edges.
+	if dir == cgraph.BackDir {
+		for i := len(obs)/2 - 1; i >= 0; i-- {
+			opp := len(obs) - 1 - i
+			obs[i], obs[opp] = obs[opp], obs[i]
+		}
+	}
+
+	// Draw the actual observations of traceflow.
+	for _, o := range obs {
+		// Construct node and edge.
+		nodeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+		node, _ := graph.CreateNode(nodeName)
+		node.SetLabel(componentString[o.Component])
+		node.SetShape(cgraph.BoxShape)
+		node.SetStyle(cgraph.RoundedNodeStyle + "," + cgraph.FilledNodeStyle)
+		node.SetFontColor("white")
+		nodes = append(nodes, node)
+		if len(nodes) > 1 {
+			edgeName := fmt.Sprintf("%s_%d", graph.Name(), len(nodes))
+			edge, _ := graph.CreateEdge(edgeName, nodes[len(nodes)-2], nodes[len(nodes)-1])
+			edge.SetDir(dir)
+			edge.SetPenWidth(3.0)
+			edge.SetColor("limegreen")
+			if len(nodes) == 2 {
+				edge.SetMinLen(1 + addNodeNum)
+			} else {
+				edge.SetMinLen(1)
+			}
+			if o.Action == v1.DROPPED && dir == cgraph.BackDir {
+				edge.SetStyle("invis")
+			}
+		}
+		// Show the name of cluster.
+		if len(o.Node) > 0 {
+			graph.SetLabel(o.Node)
+			if dir == cgraph.ForwardDir {
+				graph.SetLabelJust(cgraph.LeftJust)
+			} else {
+				graph.SetLabelJust(cgraph.RightJust)
+			}
+		}
+		// Set the pattern of node.
+		labelStr := componentString[o.Component] + "\n" + resourceString[o.Action]
+		if o.Component == v1.NETWORKPOLICY && len(o.RuleID) > 0 {
+			labelStr += "\nRuleID: " + o.RuleID
+		}
+		if o.Action == v1.DROPPED {
+			node.SetColor("violet")
+		} else {
+			node.SetColor("turquoise3")
+			if len(o.TunnelDstIP) > 0 {
+				labelStr += "\n" + o.TunnelDstIP
+			}
+		}
+		node.SetLabel(labelStr)
+	}
+	return nodes
+}
+
+func isSingleNodeTraceflow(tf *v1.Traceflow) bool {
+	// All observations are in NodeSender if a traceflow only go through one node.
+	if len(tf.Status.NodeReceiver) != 0 {
+		return false
+	}
+	return true
+}
+
+func getSrcNodeName(tf *v1.Traceflow) string {
+	if len(tf.Source.Namespace) > 0 && len(tf.Source.Pod) > 0 {
+		return tf.Source.Namespace + "/" + tf.Source.Pod
+	}
+	return ""
+}
+
+func getDstNodeName(tf *v1.Traceflow) string {
+	if len(tf.Destination.Namespace) > 0 && len(tf.Destination.Pod) > 0 {
+		return tf.Destination.Namespace + "/" + tf.Destination.Pod
+	}
+	return ""
+}
+
+func findClusterString(graphStr string, clusterName string) (startIndex int, endIndex int) {
```

Comment:
Great to see we can use this function to fix the position of source graph and destination graph! In this way, we can enable the auto refresh instead of refreshing manually.  Could you please add some comments for this function to explain the logic here?

Reply:
Comment added here.